core/src/display_object/video.rs

```diff
@@ -33,56 +33,54 @@ use swf::{DefineVideoStream, VideoCodec, VideoFrame};
 /// framerate.
 #[derive(Clone, Collect, Copy)]
 #[collect(no_drop)]
-pub struct Video<'gc>(GcCell<'gc, VideoData<'gc>>);
+pub struct Video<'gc>(Gc<'gc, VideoData<'gc>>);
 
 impl fmt::Debug for Video<'_> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         f.debug_struct("Video")
-            .field("ptr", &self.0.as_ptr())
+            .field("ptr", &Gc::as_ptr(self.0))
             .finish()
     }
 }
 
 #[derive(Clone, Collect)]
 #[collect(no_drop)]
 pub struct VideoData<'gc> {
-    base: DisplayObjectBase<'gc>,
+    base: RefLock<DisplayObjectBase<'gc>>,
 
-    avm1_text_field_bindings: Vec<Avm1TextFieldBinding<'gc>>,
+    avm1_text_field_bindings: RefLock<Vec<Avm1TextFieldBinding<'gc>>>,
 
     /// The source of the video data (e.g. an external file, a SWF bitstream)
-    source: GcCell<'gc, VideoSource<'gc>>,
```

Comment:
Have you considered removing the `Gc` here? It would avoid an allocation for non-SWF videos:
```rs
struct VideoData<'gc> {
  // ...
  source: Lock<VideoSource<'gc>>,
}

enum VideoSource<'gc> {
  Swf(Gc<'gc, SwfVideoSource>),
  // ...
}

struct SwfVideoSource {
  // ...
}
```

Reply:
Removed `Gc` entirely. Originally I thought we need to leave the reference intact but that doesn't seem to be the case.

Reply:
That seems contradicted by the doc comment of `VideoSource::Swf`:
>  NOTE: Fields within this enum will be shared across all instances of a
> particular character. If you need to mutate the video source, consider
> reallocating a new source for your specific video instead.

In particular, I think that instantiating a partially-preloaded `Video` character shouldn't make a separate copy of the frame list.

Reply:
I agree, that comments contradicts the code. What's more interesting, it was added after adding `frames`

Reply:
Well, the comment *is* correct with the current setup (`GcCell<'gc, VideoSource<'gc>>`); I'm pretty sure your change breaks the `VideoFrame` tag (because the preloading only occurs on the "master" Video character, not on any videos present on stage)

Reply:
Hmmm... are you sure about that? I'm looking at the code and I cannot see how my patch changes any behavior here.

Looking at the lifetime of the `GcCell<VideoSource>` pointer, it's allocated when creating `Video` and in `attach_netstream`. However, the pointer never escapes `Video`, so that's why I'm assuming I can just own the `VideoSource` (and thus replace a `Gc` with `RefLock`).

However, it's possible I'm not seeing something here :thinking:

Reply:
The video source can escape the current instance through`TDisplayObject::instantiate`, which creates a new `Video` object by cloning `VideoData` (and thus the `GcCell`).